api/v1alpha1/applicationset_types.go

```diff
@@ -367,6 +381,28 @@ type PullRequestGeneratorGithub struct {
 	Labels []string `json:"labels,omitempty"`
 }
 
+// PullRequestGenerator defines connection info specific to BitbucketServer.
+type PullRequestGeneratorBitbucketServer struct {
+	// Project to scan. Required.
+	Project string `json:"project"`
+	// Repo name to scan. Required.
+	Repo string `json:"repo"`
+	// The Bitbucket REST API URL to talk to e.g. https://bitbucket.org/rest Required.
+	API string `json:"api"`
+	// A regex which must match the branch name.
+	BranchMatch *string `json:"branchMatch,omitempty"`
```

Comment:
Instead of adding a `BranchMatch` filter here, could you add a `Filters []PullRequestGeneratorFilter` field to the `PullRequestGenerator` struct, to parallel the way filtering is done in `SCMProviderGenerator`? That will make things a bit more complex, but it'll make things more intuitive for the user and make sure GitHub users get filtering as well.

If you'd rather just drop this field and add filtering support on another PR, that might help get basic support in more quickly.

Reply:
If I just follow how it's done for `SCMProviderGenerator` and write it as :
```go
type PullRequestGeneratorFilter struct {
	LabelMatch *string `json:"labelMatch,omitempty"`
	BranchMatch *string `json:"branchMatch,omitempty"`
}
```
adding a label filter for Bitbucket would be valid yaml, but it isn't supported. I should make a `PullRequestGeneratorBitbucketServerFilter` struct with only the `BranchMatch` then ? I want to keep the branch regex in one way or another, there's 0 way do to filtering for Bitbucket PRs otherwise. I'm not planning on adding branch name filtering for Github PRs here though.

Btw, I didn't do it like that since the Github PR generator also has a `labels` array at the "top" level and not something like this, which would be in line with SCM provider:
```yaml
- pullRequest:
      github:
          filters:
          - labelMatch: preview
```
But I agree that it should be unified ideally.

Reply:
I'd recommend this:

```go
type PullRequestGeneratorFilter struct {
	BranchMatch *string `json:"branchMatch,omitempty"`
}
```

If someone adds github PR label filtering support later, I think the controller can throw a warning if someone adds a label filter to a bitbucketserver pull request generator. I think that missing out on apply-time validation is worth it for keeping the schema similar to the SCM generator.